### DISCUSSION TITLE
make option for littlefs

### DISCUSSION THREAD

**Discussion by permaBox** [micropython/micropython#13420] [ESP32]
Hi, I know there is an option to embed littlefs instead of fat but I can't remember.
Please light me on to make a mpy build with littlefs.

**Comment by davefes** [comment]
ESP32 are LitteFS.  Run this program to check (what_fs.py) found on the forum:
```py
"""
Dump info about a filesystem.
"""

import sys, struct

def pr(addr, label, data):
    if type(data) == bytes or type(data) == bytearray:
        data = ':'.join('%02x' % b for b in data)
    print('%04x  %-32s %s' % (addr, label, data))

def decode_bootsec_fat(b):
    print('FAT filesystem')
    pr(0, 'jump', b[0:3])
    pr(3, 'OEM name', str(b[3:11], 'ascii'))
    pr(11, 'sector size (bytes)', struct.unpack_from('<H', b, 11)[0])
    pr(13, 'cluster size (sectors)', struct.unpack_from('<B', b, 13)[0])
    pr(14, 'reserved area (sectors)', struct.unpack_from('<H', b, 14)[0])
    pr(16, 'number of FATs', struct.unpack_from('<B', b, 16)[0])
    pr(17, 'size of root dir area', struct.unpack_from('<H', b, 17)[0])
    pr(19, 'volume size (sectors)', struct.unpack_from('<H', b, 19)[0])
    pr(21, 'media descsriptor', hex(struct.unpack_from('<B', b, 21)[0]))
    pr(22, 'FAT size (sectors)', struct.unpack_from('<H', b, 22)[0])
    pr(24, 'track size (sectors)', struct.unpack_from('<H', b, 24)[0])
    pr(26, 'number of heads', struct.unpack_from('<H', b, 26)[0])
    pr(28, 'volume offset (sectors)', struct.unpack_from('<L', b, 28)[0])
    pr(32, 'volume size (32-bit) (sectors)', struct.unpack_from('<I', b, 32)[0])
    pr(36, 'physical drive number', struct.unpack_from('<B', b, 36)[0])
    pr(37, 'error flag', hex(struct.unpack_from('<B', b, 37)[0]))
    pr(38, 'extended boot signature', hex(struct.unpack_from('<B', b, 38)[0]))
    pr(39, 'volume serial number', hex(struct.unpack_from('<L', b, 39)[0]))
    pr(43, 'volume label', str(b[43:51], 'ascii'))
    pr(54, 'filesystem type', str(b[54:62], 'ascii'))
    pr(510, 'signature', hex(struct.unpack_from('<H', b, 510)[0]))


def decode_bootsec_lfs1(b):
    print("Littlefs v1 filesystem")
    pr(24, "type", struct.unpack_from("<I", b, 24)[0])
    pr(25, "elen", struct.unpack_from("<I", b, 25)[0])
    pr(26, "alen", struct.unpack_from("<I", b, 26)[0])
    pr(27, "nlen", struct.unpack_from("<I", b, 27)[0])
    pr(28, "block_size", struct.unpack_from("<I", b, 28)[0])
    pr(32, "block_count", struct.unpack_from("<I", b, 32)[0])
    pr(36, "version", hex(struct.unpack_from("<I", b, 36)[0]))
    pr(40, "magic", str(b[40:40 + 8], 'ascii'))


def decode_bootsec_lfs2(b):
    print("Littlefs v2 filesystem")
    pr(8, "magic", str(b[8:8 + 8], "ascii"))
    pr(20, "version", hex(struct.unpack_from("<I", b, 20)[0]))
    pr(24, "block_size", struct.unpack_from("<I", b, 24)[0])
    pr(28, "block_count", struct.unpack_from("<I", b, 28)[0])
    pr(32, "name_max", struct.unpack_from("<I", b, 32)[0])
    pr(36, "file_max", struct.unpack_from("<I", b, 36)[0])
    pr(40, "attr_max", struct.unpack_from("<I", b, 40)[0])


def decode_bootsec(b):
    if b[40:48] == b"littlefs":
        decode_bootsec_lfs1(b)
    elif b[8:16] == b"littlefs":
        decode_bootsec_lfs2(b)
    else:
        decode_bootsec_fat(b)

def main():
    if sys.platform == 'pyboard':
        import pyb
        #sd = pyb.SDCard()
        sd = pyb.Flash(start=0)
        bootsec = bytearray(512)
        sd.readblocks(0, bootsec)
    elif sys.platform in ('esp8266', 'esp32'):
        import esp
        bootsec = bytearray(512)
        esp.flash_read(esp.flash_user_start(), bootsec)
    else:
        with open(sys.argv[1], 'rb') as f:
            bootsec = f.read(512)
    decode_bootsec(bootsec)

main()
```

**Comment by jimmo** [comment]
Yep, ESP32 has been littlefs by default for a while now. If you have been applying updates to a board that was previously FAT then it will be still FAT, easiest way to switch is just to do a full erase-flash and re-install MicroPython.

More info at https://docs.micropython.org/en/latest/reference/filesystem.html

**Reply by gampam2000** [reply]
How or where is the filesystem defined if I compile micropython for ESP32 myself?
In all partition tables i looked there the vfs partition is always fat. For example 
https://github.com/micropython/micropython/blob/master/ports/esp32/partitions-16MiB-ota.csv

I also looked at https://github.com/micropython/micropython/blob/5d12df51fc431ecb159112bf4a40983c2ccba165/ports/esp32/modules/inisetup.py#L37C1-L46C23

There it is searched for "ffat" or "vfs". This is confusing to me, because in the partition table only "fat" and not "ffat" is written?
What would I need to put there instead of fat if I want to use litteFS as my default filesystem?

**Reply by shariltumin** [reply]
The default file system for ESP32 is Littlefs v2, no matter what you do during build. For example:

```
make V=1 BOARD=ESP32_GENERIC MICROPY_VFS_FAT=1
```

will not give you a FAT filesystem. Erasing before flashing will still give you Littlefs v2. 

If a new firmware is flashed without first performing the erase operation, the previous file system type will be retained.

This is not a problem, as you can change the file system to whatever you want after the system boots. Littlefs v1 is not supported. All files will be deleted. Best to do a backup first.

```
import vfs
# change to FAT
vfs.umount('/')
vfs.VfsFat.mkfs(bdev)
vfs.mount(bdev, '/')
```

or

```
import vfs
# change to LFS
vfs.umount('/')
vfs.VfsLfs2.mkfs(bdev)
vfs.mount(bdev, '/')
```

The "vfs" in the partition table are configuration values for offset and size to esp-idf during build. The important label is "vfs".

```
>>> bdev.info()
(1, 129, 2097152, 2097152, 'vfs', False)
>>> 
```

The important thing for creating a filesystem is the "vfs" partition. The flashbdev.py and inisetup.py look for either 'vfs' or 'ffat'. If the 'Name' is set to 'ffat' then you will get a FAT filesystem. You can call it whatever you want, but you will not get a filesystem. You have to create it after boot at REPL.

**Reply by gampam2000** [reply]
ah ok, so if I write "ffat" into the partition table instead of vfs i would get a fat filesystem?

**Reply by shariltumin** [reply]
Yes. You may need to erase the flash before reflashing the MP firmware.

**Reply by gampam2000** [reply]
thanks for the clarification. The documentation https://docs.micropython.org/en/latest/reference/filesystem.html#working-with-filesystems could be clearer about this.
